java/arcs/android/storage/service/BindingContext.kt

```diff
@@ -74,81 +74,93 @@ class BindingContext(
     @VisibleForTesting
     val id = nextId.incrementAndGet()
 
-    /** The local [CoroutineContext]. */
+    /**
+     * The local [CoroutineContext], and a [CoroutineScope] that wraps it.
+     *
+     * All of the implemented AIDL methods will be immediately dispatched using this scope.
+     */
     private val job = Job(parentCoroutineContext[Job])
     private val coroutineContext =
         parentCoroutineContext + job + CoroutineName("BindingContext-$id")
+    private val scope = CoroutineScope(coroutineContext)
 
     override fun idle(timeoutMillis: Long, resultCallback: IResultCallback) {
-        bindingContextStatisticsSink.traceTransaction("idle") {
-            bindingContextStatisticsSink.measure(coroutineContext) {
-                try {
-                    withTimeout(timeoutMillis) {
-                        store().idle()
+        scope.launch {
+            bindingContextStatisticsSink.traceTransaction("idle") {
+                bindingContextStatisticsSink.measure {
+                    try {
+                        withTimeout(timeoutMillis) {
+                            store().idle()
+                        }
+                        resultCallback.onResult(null)
+                    } catch (e: Throwable) {
+                        resultCallback.onResult(
+                            CrdtException("Exception occurred while awaiting idle", e).toProto()
+                                .toByteArray()
+                        )
                     }
-                    resultCallback.onResult(null)
-                } catch (e: Throwable) {
-                    resultCallback.onResult(
-                        CrdtException("Exception occurred while awaiting idle", e).toProto()
-                            .toByteArray()
-                    )
                 }
             }
         }
     }
 
     @Suppress("UNCHECKED_CAST")
     override fun registerCallback(callback: IStorageServiceCallback): Int {
-        var callbackToken = 0
-        bindingContextStatisticsSink.traceTransaction("registerCallback") {
-            val proxyCallback = ProxyCallback<CrdtData, CrdtOperation, Any?> { message ->
-                // Asynchronously pass the message along to the callback. Use a supervisorScope here
-                // so that we catch any exceptions thrown within and re-throw on the same coroutine
-                // as the callback-caller.
-                supervisorScope {
-                    callback.onProxyMessage(message.toProto().toByteArray())
+        // TODO(b/162946893) - Make this oneway and add a result callback
```

Comment:
This is the right thing to do!

`scope.launch` + `CompletableFuture` is a bit dangerous if `ServiceStore` and `BindingContext` are run at the same thread. (currently it's safe in GH and G3).
`CompletableFuture.get()` blocks the current thread so the coroutine generated by `scope.launch` may not be run for good.

Originally runBlocking creates an event loop and returns only when its inner/child jobs/coroutines are completed.

Reply:
I see, that's a good point. I'll convert this back to using `runBlocking` for now, instead of the completable future.

Reply:
Actually, since this is ok *for now* I will just merge this soon, and then quickly create a follow-up with the service interface changes described here.